File: twakkal_lwp/src/main/java/com/yassin/sfax/tawakkolalaallah/LiveWallpaper.java
package com.yassin.sfax.tawakkolalaallah;

import java.util.ArrayList;

import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.service.wallpaper.WallpaperService;
import android.view.MotionEvent;
import android.view.SurfaceHolder;

import com.sfaxdroid.base.AnimatedItem;

public class LiveWallpaper extends WallpaperService {

    @Override
    public Engine onCreateEngine() {
        return new WallpaperPatternEngine();
    }

    class WallpaperPatternEngine extends Engine {
        private final Handler mHandler = new Handler();
        private float mTouchX = -1;
        private float mTouchY = -1;
        private final Runnable mDrawPattern = new Runnable() {
            public void run() {
                drawFrame();
            }
        };
        private final ArrayList<AnimatedItem> animatedItems;
        private int newCloudIteration = 0;
        private Bitmap background, rainingItem, touchingItem;
        final SurfaceHolder surfaceHolder = getSurfaceHolder();
        private boolean mVisible;

        WallpaperPatternEngine() {
            final Paint paint = new Paint();
            paint.setColor(0xffffffff);
            paint.setAntiAlias(true);
            paint.setStrokeWidth(2);
            paint.setStrokeCap(Paint.Cap.ROUND);
            paint.setStyle(Paint.Style.STROKE);
            animatedItems = new ArrayList<>();
            animatedItems.add(new AnimatedItem(10, 50));
            background = BitmapFactory.decodeResource(getResources(),
                    R.drawable.wallpaper1);
            rainingItem = BitmapFactory.decodeResource(getResources(),
                    R.drawable.tw41);
            touchingItem = BitmapFactory.decodeResource(getResources(),
                    R.drawable.tw41);
        }

        @Override
        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
            setTouchEventsEnabled(true);
        }

        @Override
        public void onDestroy() {
            super.onDestroy();
            mHandler.removeCallbacks(mDrawPattern);
        }

        @Override
        public void onVisibilityChanged(boolean visible) {

            SharedPreferences sharedPrefs = PreferenceManager
                    .getDefaultSharedPreferences(LiveWallpaper.this);

            boolean Anim1 = sharedPrefs.getBoolean("checkBox1", true);
            boolean Anim2 = sharedPrefs.getBoolean("checkBox2", true);

            String syncFreq = sharedPrefs.getString("prefSyncFrequency", "NULL");

            background = ResourceUtils.Companion.getBackground(getBaseContext(), syncFreq);
            rainingItem = ResourceUtils.Companion.getRainingItem(getBaseContext(), syncFreq, Anim1);
            touchingItem = ResourceUtils.Companion.getTouchingItem(getBaseContext(), syncFreq, Anim2);

            mVisible = visible;
            if (visible) {
                drawFrame();
            } else {
                mHandler.removeCallbacks(mDrawPattern);
            }
        }

        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format,
                                     int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);
            drawFrame();
        }

        @Override
        public void onSurfaceCreated(SurfaceHolder holder) {
            super.onSurfaceCreated(holder);
        }

        @Override
        public void onSurfaceDestroyed(SurfaceHolder holder) {
            super.onSurfaceDestroyed(holder);
            mVisible = false;
            mHandler.removeCallbacks(mDrawPattern);
        }

        @Override
        public void onOffsetsChanged(float xOffset, float yOffset, float xStep,
                                     float yStep, int xPixels, int yPixels) {
            drawFrame();
        }

        @Override
        public void onTouchEvent(MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_MOVE) {
                mTouchX = event.getX();
                mTouchY = event.getY();
            }
            super.onTouchEvent(event);
        }

        void drawFrame() {
            Canvas canvas = null;
            try {

                canvas = surfaceHolder.lockCanvas();
                if (canvas != null) {
                    drawPattern(canvas);
                }
            } finally {
                if (canvas != null)
                    surfaceHolder.unlockCanvasAndPost(canvas);
            }

            mHandler.removeCallbacks(mDrawPattern);
            if (mVisible) {
                mHandler.postDelayed(mDrawPattern, 1000 / 25);
            }
        }

        void drawPattern(Canvas canvas) {
            canvas.save();
            drawBackground(canvas);
            drawRainingItem(canvas);
            drawTouchingItem(canvas);
        }

        private void drawBackground(Canvas canvas) {
            canvas.drawColor(0xff000000);
            canvas.drawBitmap(background, 0, 0, null);
        }

        private void drawRainingItem(Canvas canvas) {
            int screenWidth = canvas.getWidth();
            if (newCloudIteration > 200 + (int) (Math.random() * 1000 % 100)) {
                animatedItems.add(new AnimatedItem(
                        (int) (Math.random() * 1000 % screenWidth), 0));
                newCloudIteration = 0;
            }
            newCloudIteration++;
            int screenHeight = canvas.getHeight();
            for (int i = 0; i < animatedItems.size(); i++) {
                canvas.drawBitmap(rainingItem, animatedItems.get(i).x - rainingItem.getWidth() / 2,
                        animatedItems.get(i).y - rainingItem.getHeight() / 2, null);
                animatedItems.get(i).y++;
                if (animatedItems.get(i).y > screenHeight) {
                    animatedItems.remove(i);
                    i--;
                }
            }
        }

        private void drawTouchingItem(Canvas canvas) {
            if (mTouchX >= 0 && mTouchY >= 0) {
                canvas.drawBitmap(touchingItem, mTouchX - touchingItem.getWidth() / 2, mTouchY
                        - touchingItem.getHeight() / 2, null);
                canvas.restore();
            }
        }


    }
}

File: mini-app-raining-anim/src/main/java/com/yassin/sfax/tawakkolalaallah/LiveWallpaper.kt
package com.yassin.sfax.tawakkolalaallah

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.os.Handler
import android.service.wallpaper.WallpaperService
import android.view.MotionEvent
import android.view.SurfaceHolder
import com.sfaxdroid.mini.base.AnimatedItem
import com.sfaxdroid.mini.base.BaseConstants
import com.yassin.sfax.tawakkolalaallah.ResourceUtils.Companion.getBackground
import com.yassin.sfax.tawakkolalaallah.ResourceUtils.Companion.getRainingItem
import com.yassin.sfax.tawakkolalaallah.ResourceUtils.Companion.getTouchingItem
import java.util.*

class LiveWallpaper : WallpaperService() {

    override fun onCreateEngine(): Engine {
        return WallpaperPatternEngine()
    }

    internal inner class WallpaperPatternEngine :
        Engine() {
        private val mHandler = Handler()
        private var mTouchX = -1f
        private var mTouchY = -1f
        private val mDrawPattern = Runnable { drawFrame() }
        private val animatedItems: ArrayList<AnimatedItem> = ArrayList()
        private var newCloudIteration = 0

        private var background = BitmapFactory.decodeResource(
            resources,
            R.drawable.wallpaper1
        )

        private var rainingItem = BitmapFactory.decodeResource(
            resources,
            R.drawable.style1_raining
        )

        private var touchingItem = BitmapFactory.decodeResource(
            resources,
            R.drawable.style1_raining
        )

        private var mVisible = false

        init {
            animatedItems.add(AnimatedItem(10, 50))
        }

        override fun onCreate(surfaceHolder: SurfaceHolder) {
            super.onCreate(surfaceHolder)
            setTouchEventsEnabled(true)
        }

        override fun onDestroy() {
            super.onDestroy()
            mHandler.removeCallbacks(mDrawPattern)
        }

        override fun onVisibilityChanged(visible: Boolean) {
            val sharedPrefs = getSharedPreferences(BaseConstants.PREF_NAME, Context.MODE_PRIVATE)
            val rainingPref = sharedPrefs.getBoolean("rainingPref", true)
            val touchingPref = sharedPrefs.getBoolean("touchingPref", true)
            val themePref = sharedPrefs.getString("themePref", "Style 1")

            background = getBackground(
                baseContext,
                themePref!!
            )
            rainingItem = getRainingItem(
                baseContext,
                themePref,
                rainingPref
            )
            touchingItem = getTouchingItem(
                baseContext,
                themePref,
                touchingPref
            )
            mVisible = visible
            if (visible) {
                drawFrame()
            } else {
                mHandler.removeCallbacks(mDrawPattern)
            }
        }

        override fun onSurfaceChanged(
            holder: SurfaceHolder, format: Int,
            width: Int, height: Int
        ) {
            super.onSurfaceChanged(holder, format, width, height)
            drawFrame()
        }

        override fun onSurfaceDestroyed(holder: SurfaceHolder) {
            super.onSurfaceDestroyed(holder)
            mVisible = false
            mHandler.removeCallbacks(mDrawPattern)
        }

        override fun onOffsetsChanged(
            xOffset: Float, yOffset: Float, xStep: Float,
            yStep: Float, xPixels: Int, yPixels: Int
        ) {
            drawFrame()
        }

        override fun onTouchEvent(event: MotionEvent) {
            if (event.action == MotionEvent.ACTION_MOVE) {
                mTouchX = event.x
                mTouchY = event.y
            }
            super.onTouchEvent(event)
        }

        private fun drawFrame() {
            var canvas: Canvas? = null
            try {
                canvas = surfaceHolder.lockCanvas()
                canvas?.let { drawPattern(it) }
            } finally {
                canvas?.let { surfaceHolder.unlockCanvasAndPost(canvas) }
            }
            mHandler.removeCallbacks(mDrawPattern)
            if (mVisible) {
                mHandler.postDelayed(mDrawPattern, 1000 / 25.toLong())
            }
        }

        private fun drawPattern(canvas: Canvas) {
            canvas.save()
            drawBackground(canvas)
            drawRainingItem(canvas)
            drawTouchingItem(canvas)
        }

        private fun drawBackground(canvas: Canvas) {
            canvas.drawColor(-0x1000000)
            canvas.drawBitmap(background, 0f, 0f, null)
        }

        private fun drawRainingItem(canvas: Canvas) {
            val screenWidth = canvas.width
            if (newCloudIteration > 200 + (Math.random() * 1000 % 100).toInt()) {
                animatedItems.add(
                    AnimatedItem(
                        (Math.random() * 1000 % screenWidth).toInt(), 0
                    )
                )
                newCloudIteration = 0
            }
            newCloudIteration++
            var i = 0
            while (i < animatedItems.size) {
                canvas.drawBitmap(
                    rainingItem, animatedItems[i].x - rainingItem.width / 2.toFloat(),
                    animatedItems[i].y - rainingItem.height / 2.toFloat(), null
                )
                animatedItems[i].y++
                if (animatedItems[i].y > canvas.height) {
                    animatedItems.removeAt(i)
                    i--
                }
                i++
            }
        }

        private fun drawTouchingItem(canvas: Canvas) {
            if (mTouchX >= 0 && mTouchY >= 0) {
                canvas.drawBitmap(
                    touchingItem, mTouchX - touchingItem.width / 2, mTouchY
                            - touchingItem.height / 2, null
                )
                canvas.restore()
            }
        }

    }
}